ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/block/AbstractBlockProcessor.java

```diff
@@ -793,8 +795,29 @@ public void processVoluntaryExits(
   protected void processVoluntaryExitsNoValidation(
       MutableBeaconState state, SszList<SignedVoluntaryExit> exits)
       throws BlockProcessingException {
+    // long start = System.currentTimeMillis();
+
     safelyProcess(
         () -> {
+          if (exits.isEmpty()) {
+            return;
+          }
+
+          AtomicReference<UInt64> maxEpoch = new AtomicReference<>(UInt64.ZERO);
+
+          List<Validator> exitedValidators =
+              state.getValidators().stream()
+                  .filter(validator -> !validator.getExitEpoch().equals(FAR_FUTURE_EPOCH))
+                  .peek(validator -> maxEpoch.set(maxEpoch.get().max(validator.getExitEpoch())))
+                  .collect(toList());
```

Comment:
i'll look at this more closely in the morning... I think I agree with the direction, just some of the details I think we need to be careful with, esp like here where we're passing lists constantly, that'll be a big memory impact potentially.... this list will get bigger for ever...

Reply:
During the we I thought about an additional improvement and a way to make it cleaner..

Reply:
Not sure we can avoid to collect that list. Take a look to my last version